src/executor/evaluate/evaluated.rs

```diff
@@ -343,6 +346,23 @@ impl<'a> Evaluated<'a> {
             StringRef(_) => unreachable(),
         }
     }
+
+    pub fn cast(&self, data_type: &DataType) -> Result<Evaluated<'a>> {
+        use Evaluated::*;
+
+        let cast_literal =
+            |value: &AstValue| cast_ast_value(value.to_owned(), data_type).map(Evaluated::Literal);
+        let cast_value = |value: &data::Value| value.cast(data_type).map(Evaluated::Value);
+
+        // Decided: Due to the explicit call, we can abandon -Ref
+        match self {
+            StringRef(value) => cast_literal(&AstValue::SingleQuotedString(value.to_string())),
+            LiteralRef(value) => cast_literal(value.to_owned()),
+            Literal(value) => cast_literal(value),
+            ValueRef(value) => cast_value(value.to_owned()),
+            Value(value) => cast_value(value),
+        }
+    }
```

Comment:
You can simply take ownership of the result
```rust
pub fn cast(self, data_type: &DataType) -> Result<Evaluated<'a>> {
    use Evaluated::*;

    let cast_literal =
        |value: &AstValue| cast_ast_value(value, data_type).map(Evaluated::Literal);
    let cast_value = |value: &data::Value| value.cast(data_type).map(Evaluated::Value);

    match self {
        StringRef(value) => cast_literal(&AstValue::SingleQuotedString(value.to_string())),
        LiteralRef(value) => cast_literal(value),
        Literal(value) => cast_literal(&value),
        ValueRef(value) => cast_value(value),
        Value(value) => cast_value(&value),
    }
}
```

Then 
`pub fn cast_ast_value` would be better to take `&AstValue` like `data::Value -> cast` method.
```rust
pub fn cast_ast_value(value: &AstValue, data_type: &DataType) -> Result<AstValue> {
    match (data_type, value) {
        (DataType::Boolean, AstValue::SingleQuotedString(value))
        | (DataType::Boolean, AstValue::Number(value)) => Ok(match value.to_uppercase().as_str() {
            "TRUE" | "1" => Ok(AstValue::Boolean(true)),
            "FALSE" | "0" => Ok(AstValue::Boolean(false)),
            _ => Err(EvaluateError::ImpossibleCast),
        }?),
        (DataType::Int, AstValue::Number(value)) => Ok(AstValue::Number(
            value
                .parse::<f64>()
                .map_err(|_| EvaluateError::UnreachableImpossibleCast)?
                .trunc()
                .to_string(),
        )),
        (DataType::Int, AstValue::SingleQuotedString(value))
        | (DataType::Float(_), AstValue::SingleQuotedString(value)) => Ok(AstValue::Number(value.clone())),
        (DataType::Int, AstValue::Boolean(value))
        | (DataType::Float(_), AstValue::Boolean(value)) => Ok(AstValue::Number(
            (if *value { "1" } else { "0" }).to_string(),
        )),
        (DataType::Float(_), AstValue::Number(value)) => Ok(AstValue::Number(value.clone())),
        (DataType::Text, AstValue::Boolean(value)) => Ok(AstValue::SingleQuotedString(
            (if *value { "TRUE" } else { "FALSE" }).to_string(),
        )),
        (DataType::Text, AstValue::Number(value)) => Ok(AstValue::SingleQuotedString(value.clone())),
        (_, AstValue::Null) => Ok(AstValue::Null),
        _ => Err(EvaluateError::UnimplementedCast.into()),
    }
}
```

Reply:
fyi. That is for reducing `clone()` calls for `CAST`, using this way, it can handle `cast` with a single clone call.

Reply:
Ah! Thanks!